src/compose/ComposeBox.js

```diff
@@ -153,270 +136,352 @@ const updateTextInput = (textInput, text) => {
+  // But if callers do that, they should *also* update this state, because
+  // it's our best read of what the actual value is, so it needs to be
+  // up-to-date. That's what setTopicInputValue (far below) is for; it does
+  // both.
+  // TODO(?): Make custom Hook to encapsulate this uncontrolled-input logic.
+  const [messageInputState, setMessageInputState] = useState<{|
+    value: string,
+    selection: InputSelection,
+  |}>({
+    value: props.initialMessage ?? '',
     selection: { start: 0, end: 0 },
-    numUploading: 0,
-  };
+  });
 
-  componentWillUnmount() {
-    clearTimeout(this.inputBlurTimeoutId);
-    this.inputBlurTimeoutId = null;
-  }
+  useEffect(
+    () => () => {
+      clearTimeout(inputBlurTimeoutId.current);
+      inputBlurTimeoutId.current = null;
+    },
+    [],
+  );
+
+  const prevMessageInputState = usePrevious(messageInputState);
+  useEffect(() => {
+    const { dispatch, isEditing, narrow } = props;
+    const messageInputValue = messageInputState.value;
+    const prevMessageInputValue = prevMessageInputState?.value;
+
+    if (prevMessageInputValue !== messageInputValue) {
+      if (messageInputValue.length === 0) {
+        dispatch(sendTypingStop(narrow));
+      } else {
+        dispatch(sendTypingStart(narrow));
+      }
+      if (!isEditing) {
+        dispatch(draftUpdate(narrow, messageInputValue));
+      }
+    }
+  }, [props, messageInputState, prevMessageInputState]);
 
-  updateIsFocused = () => {
-    this.setState(state => ({
-      ...state,
-      isFocused: state.isMessageFocused || state.isTopicFocused,
-    }));
-  };
+  const updateIsFocused = useCallback(() => {
+    setFocusState(s => ({ ...s, either: s.message || s.topic }));
+  }, [setFocusState]);
 
-  getCanSelectTopic = () => {
-    const { isEditing, narrow } = this.props;
+  const getCanSelectTopic = useCallback(() => {
+    const { isEditing, narrow } = props;
     if (isEditing) {
       return isStreamOrTopicNarrow(narrow);
     }
     if (!isStreamNarrow(narrow)) {
       return false;
     }
-    return this.state.isFocused;
-  };
-
-  setMessageInputValue = (message: string) => {
-    updateTextInput(this.messageInputRef.current, message);
-    this.handleMessageChange(message);
-  };
-
-  setTopicInputValue = (topic: string) => {
-    updateTextInput(this.topicInputRef.current, topic);
-    this.handleTopicChange(topic);
-  };
-
-  insertMessageTextAtCursorPosition = (text: string) => {
-    const { message, selection } = this.state;
+    return focusState.either;
+  }, [props, focusState.either]);
 
-    this.setMessageInputValue(
-      message.slice(0, selection.start) + text + message.slice(selection.end, message.length),
-    );
-  };
+  const handleMessageChange = useCallback(
+    (newValue: string) => {
+      setMessageInputState(s => ({ ...s, value: newValue }));
+      setIsMenuExpanded(false);
+    },
+    [setMessageInputState, setIsMenuExpanded],
```

Comment:
These can be left out of the dependency list -- the ESLint rule is smart enough to see that because they're setters from `useState`, they never change. (Similarly in some others below.)